pkg/cmd/server/origin/audit.go

```diff
@@ -0,0 +1,40 @@
+package origin
+
+import (
+	"net/http"
+
+	"github.com/golang/glog"
+	"github.com/pborman/uuid"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+)
+
+type auditResponseWriter struct {
+	http.ResponseWriter
+	id string
+}
+
+func (a *auditResponseWriter) WriteHeader(code int) {
+	glog.Infof("AUDIT: %s: response:%d", a.id, code)
+	a.ResponseWriter.WriteHeader(code)
+}
+
+func (c *MasterConfig) auditHandler(handler http.Handler) http.Handler {
+	if !c.Options.AuditConfig.Enabled {
+		return handler
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx, _ := c.RequestContextMapper.Get(req)
+		user, _ := kapi.UserFrom(ctx)
+		namespace := kapi.NamespaceValue(ctx)
+		if len(namespace) == 0 {
+			namespace = "<none>"
+		}
+		id := uuid.NewRandom().String()
+
+		glog.Infof("AUDIT: %s: ip:%s, method:%s, username:%s, namespace: %s, uri:%v",
```

Comment:
You either need to add an AUDIT message in the impersonation filter using the same UID (obviously you'd have to attach it to the context) or this needs to understand the impersonation annotation.